libsmartcols/samples/filter.c

```diff
@@ -0,0 +1,217 @@
+ * GNU Lesser General Public License.
+ */
+#include <stdlib.h>
+#include <unistd.h>
+#include <string.h>
+#include <errno.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <dirent.h>
+#include <getopt.h>
+
+#include "c.h"
+#include "nls.h"
+#include "strutils.h"
+#include "xalloc.h"
+
+#include "libsmartcols.h"
+
+
+enum { COL_NAME, COL_NUM, COL_FLOAT, COL_STRING };
+
+/* add columns to the @tb */
+static void setup_columns(struct libscols_table *tb)
+{
+	struct libscols_column *col;
+
+	col = scols_table_new_column(tb, "NAME", 0, 0);
+	if (!col)
+		goto fail;
+	scols_column_set_json_type(col, SCOLS_JSON_STRING);
+
+	col = scols_table_new_column(tb, "NUM", 0, 0);
+	if (!col)
+		goto fail;
+	scols_column_set_json_type(col, SCOLS_JSON_NUMBER);
+
+	col = scols_table_new_column(tb, "FLOAT", 0, 0);
+	if (!col)
+		goto fail;
+	scols_column_set_json_type(col, SCOLS_JSON_FLOAT);
+
+	col = scols_table_new_column(tb, "STRING", 0, 0);
+	if (!col)
+		goto fail;
+	scols_column_set_json_type(col, SCOLS_JSON_STRING);
+
+	return;
+fail:
+	scols_unref_table(tb);
+	err(EXIT_FAILURE, "failed to create output columns");
+}
+
+static struct libscols_line *add_line(struct libscols_table *tb, int n, int empty)
+{
+	char *data = NULL;
+	struct libscols_line *ln = scols_table_new_line(tb, NULL);
+	if (!ln)
+		err(EXIT_FAILURE, "failed to create output line");
+
+	if (empty != 1) {
+		xasprintf(&data, "#%d", n);
+		if (scols_line_refer_data(ln, COL_NAME, data))
+			goto fail;
+	}
+	if (empty != 2) {
+		xasprintf(&data, "%d", n);
+		if (scols_line_refer_data(ln, COL_NUM, data))
+			goto fail;
+	}
+	if (empty != 3) {
+		xasprintf(&data, "%d.%d", n, n);
+		if (scols_line_refer_data(ln, COL_FLOAT, data))
+			goto fail;
+	}
+	if (empty != 4) {
+		xasprintf(&data, "str%dstr", n);
+		if (scols_line_refer_data(ln, COL_STRING, data))
+			goto fail;
+	}
+	return ln;
+fail:
+	scols_unref_table(tb);
+	err(EXIT_FAILURE, "failed to create output line");
+}
+
+static struct libscols_filter *init_filter(struct libscols_table *tb, const char *query, int dump)
+{
+	struct libscols_iter *itr;
+	struct libscols_filter *f = scols_new_filter(NULL);
+	const char *name = NULL;
+	int rc = 0;
+
+	if (!f)
+		err(EXIT_FAILURE, "failed to allocate filter");
+	if (scols_filter_parse_string(f, query) != 0)
+		errx(EXIT_FAILURE, "failed to parse filter: %s",
+				scols_filter_get_errmsg(f));
+
+	itr = scols_new_iter(SCOLS_ITER_FORWARD);
+	if (!itr)
+		err(EXIT_FAILURE, "failed to allocate iterator");
+
+	while (scols_filter_next_holder(f, itr, &name, 0) == 0) {
+		struct libscols_column *col;
+
+		col = scols_table_get_column_by_name(tb, name);
+		if (!col) {
+			warnx("unknown column '%s' in filter", name);
+			rc++;
+			continue;
+		}
+		scols_filter_assign_column(f, itr, name, col);
+	}
+
+	scols_free_iter(itr);
+	if (dump)
+		scols_dump_filter(f, stdout);
+	if (rc) {
+		scols_unref_filter(f);
+		f = NULL;
+	}
+	return f;
+}
+
+int main(int argc, char *argv[])
```

Comment:
Would be nice if this could be integrated into the CI, or some other dedicated testcase.

Also, would it make sense to fuzz the filter evaluation?